rskj-core/src/test/java/org/ethereum/vm/program/ProgramResultTest.java

```diff
@@ -0,0 +1,106 @@
+import org.ethereum.vm.PrecompiledContracts;
+import org.junit.Assert;
+import org.junit.Test;
+
+public class ProgramResultTest {
+
+    private final TestSystemProperties config = new TestSystemProperties();
+
+    @Test
+    public void add_internal_tx_one_level() {
+        Transaction originTx = getOriginTransaction();
+        ProgramResult programResult = new ProgramResult();
+        InternalTransaction internalTx = programResult.addInternalTransaction(
+            originTx,
+            0,
+            DataWord.ONE.getByteArrayForStorage(),
+            DataWord.ONE,
+            DataWord.ONE,
+            new ECKey().getAddress(),
+            new ECKey().getAddress(),
+            new byte[] {},
+            new byte[] {},
+            ""
+        );
+
+        Assert.assertArrayEquals(originTx.getHash().getBytes(), internalTx.getOriginHash());
+        Assert.assertArrayEquals(originTx.getHash().getBytes(), internalTx.getParentHash());
+    }
+
+    @Test
+    public void add_interenal_tx_two_levels() {
+        Transaction originTx = getOriginTransaction();
+        ProgramResult programResult = new ProgramResult();
+        InternalTransaction internalTx1 = programResult.addInternalTransaction(
+            originTx,
+            0,
+            DataWord.ONE.getByteArrayForStorage(),
+            DataWord.ONE,
+            DataWord.ONE,
+            new ECKey().getAddress(),
+            new ECKey().getAddress(),
+            new byte[] {},
+            new byte[] {},
+            ""
+        );
+        InternalTransaction internalTx2 = programResult.addInternalTransaction(
+            internalTx1,
+            0,
+            DataWord.ONE.getByteArrayForStorage(),
+            DataWord.ONE,
+            DataWord.ONE,
+            new ECKey().getAddress(),
+            new ECKey().getAddress(),
+            new byte[] {},
+            new byte[] {},
+            ""
+        );
+
+        Assert.assertArrayEquals(originTx.getHash().getBytes(), internalTx2.getOriginHash());
+        Assert.assertArrayEquals(internalTx1.getHash().getBytes(), internalTx2.getParentHash());
+    }
+
+    @Test
+    public void add_interenal_tx_many_levels() {
+        Transaction originTx = getOriginTransaction();
+        ProgramResult programResult = new ProgramResult();
+
+        Transaction internalTxN = originTx;
+        for (int i = 0; i < 3; i++) {
+            internalTxN = programResult.addInternalTransaction(
+                internalTxN,
+                0,
+                DataWord.ONE.getByteArrayForStorage(),
+                DataWord.ONE,
+                DataWord.ONE,
+                new ECKey().getAddress(),
+                new ECKey().getAddress(),
+                new byte[] {},
+                new byte[] {},
+                ""
+            );
+        }
+        InternalTransaction result = (InternalTransaction)internalTxN;
+
+        Assert.assertArrayEquals(originTx.getHash().getBytes(), result.getOriginHash());
```

Comment:
Since multiple levels are being tested, perhaps an assert could be added that identifies that the parent hash is not the origin hash in this case.